lib/Doctrine/ORM/Mapping/ComponentMetadata.php

```diff
@@ -154,7 +154,8 @@ public function hasProperty(string $propertyName) : bool
     }
 
     /**
-     * @return \ArrayIterator|ColumnMetadata[]
+     * @return \ArrayIterator
+     * @psalm-return \ArrayIterator<ColumnMetadata>
```

Comment:
What was the problem here? `\ArrayIterator|ColumnMetadata[]` says _\ArrayIterator that contains ColumnMetadatas_. Wouldn't add any vendor-specific hints with non-standard syntax if possible. Unfortunately generic-like syntax (`type<K, V>`) is not well supported by SA tools and IDEs so it's not used even on places where it makes more sense (hashes vs. lists). :(

Reply:
> `\ArrayIterator|ColumnMetadata[]` says *\ArrayIterator that contains ColumnMetadatas*

As I understand union types, `A|B` means either `A` or `B` are acceptable values. So that implies that an `array` is an acceptable value, which is not the case here.

The type given is just what PHPStorm needs to have an idea about what's being iterated over, and that's fairly vendor-specific, just not explicitly so.

Reply:
Yes, you're right. But unfortunately, due to limitation of current IDEs and SA/CS tools, it's not currently possible to express these properly. So the least troublesome variant is _abusing_ union syntax for now. I really wish this gets improved in future, we already see increasing adoption of Psalm and PhpStorm which leverage better type systems, let's hope PhpStorm will not be ignorant in this manner and improve too. 👍 
So far, generics syntax is not well supported and intersection syntax is completely unsupported (i.e. PhpStorm fails to interpret both).

Reply:
> due to limitation of current IDEs and SA/CS tools

*some* SA/CS tools...

> So the least troublesome variant is abusing union syntax for now

I'm not sure that follows. Surely the least troubling variant is just something like
```
 @return \ArrayIterator of ColumnMetadata
```

Nothing is violated there.

Reply:
I think the original return type should be kept, at least until we decide to start using intersections - which won't likely happen before being supported in IDEs.

Reply:
`ArrayIterator&ColumnMetadata[]` isn't a valid intersection type either (as an array can never be an `ArrayIterator` object.

But anyway, I'll revert this change and [add support for this annotation to Psalm](https://github.com/vimeo/psalm/issues/479) – if nothing else, this back and forth tells me that PHPStorm users really like this convention.